### DISCUSSION TITLE
How to check if all references "under" a node are resolved and parsed correctly (validators go crazy)

### DISCUSSION THREAD

**Discussion by goto40** [eclipse-langium/langium#937] [Q&A] [answered]
I want to check if all references under a node are resolved. I need this to prevent running into assertions while validating a model.

Is the following the right way of doing this?
```ts
streamReferences(node).reduce((ok, description)=>ok && (description.reference.ref!==undefined), true);
```

**Comment by msujew** [comment, marked as answer]
Hey @goto40,

Do you need *all* (as in: even those in children) references or only those defined on the specified node? If those on the specified node are enough, then your code should do well for that task 👍 

Otherwise, you will need to wrap this with a call to `streamAst(node)`.

**Reply by goto40** [reply]
Thank you @msujew for the clarification (and the prompt response!): I was not aware of this detail.

Is there also a way to check if the whole model was parsed correctly (not only if the references are resolved)? While entering a model I get `undefined` for model attrbutes which normally are always defined (because the parse is incomplete). My validators then go crazy... Is there a way to deal with this? (apart from checking every single field before accessing it).

**Reply by msujew** [reply]
> Is there also a way to check if the whole model was parsed correctly (not only if the references are resolved)?

You can always look at the `parseResult` of a document and see whether the `parserErrors.length === 0` check holds.

> Is there a way to deal with this? (apart from checking every single field before accessing it).

We always had the same problem in Xtext. I don't really think there's a better way to deal with this than being very careful in what you access. Generally, you only undefined checks in validation, scoping and LSP code, but that's already quite a lot. So unfortunately no easy way to deal with this, at least none that I've found.

**Reply by goto40** [reply]
Thank you very much again for the promp response and the advice / for sharing your experience!
